src/smt/assertions.cpp

```diff
@@ -122,10 +124,65 @@ void Assertions::addFormula(TNode n,
     // if a non-recursive define-fun, just add as a top-level substitution
     if (n.getKind() == Kind::EQUAL && n[0].isVar())
     {
-      // A define-fun is an assumption in the overall proof, thus
-      // we justify the substitution with ASSUME here.
+      Trace("smt-define-fun")
+          << "Define fun: " << n[0] << " = " << n[1] << std::endl;
+      bool isLambda = n[1].getKind() == Kind::LAMBDA;
+      NodeManager* nm = nodeManager();
+      TrustSubstitutionMap& tsm = d_env.getTopLevelSubstitutions();
+      // If it is a lambda, we rewrite the body, otherwise we rewrite itself.
+      // For lambdas, we prefer rewriting only the body since we don't want
+      // higher-order rewrites (e.g. value normalization) to apply by default.
+      TrustNode defRewBody =
+          tsm.applyTrusted(isLambda ? n[1][1] : n[1], d_env.getRewriter());
+      Node defRew = n[1];
+      if (!defRewBody.isNull())
+      {
+        defRew = defRewBody.getNode();
+        defRew = isLambda ? nm->mkNode(Kind::LAMBDA, n[1][0], defRew) : defRew;
+      }
+      // if we need to track proofs
+      if (d_env.isProofProducing())
+      {
+        // initialize the proof generator if not already done so
+        if (d_defFunRewPf == nullptr)
+        {
+          d_defFunRewPf = std::make_shared<LazyCDProof>(d_env);
+        }
+        if (defRew != n[1])
```

Comment:
you could lift this test so you don't need to initialize the proof generator if you're not gonna use it now.

Reply:
Actually, this is not the case, we always have to at least say that it is an assumption, I'll reorg the code block to make this clearer.